src/models/src/entity/atproto.rs

```diff
@@ -0,0 +1,368 @@
+
+                        if link.user_id != user.id {
+                            return Err(ErrorResponse::new(
+                                ErrorResponseType::BadRequest,
+                                "bad user_id in link cookie".to_string(),
+                            ));
+                        }
+
+                        if link.user_email != user.email {
+                            return Err(ErrorResponse::new(
+                                ErrorResponseType::BadRequest,
+                                "Invalid E-Mail".to_string(),
+                            ));
+                        }
+
+                        user.auth_provider_id = Some("atproto".to_owned());
+                        user.federation_uid = Some(claims_user_id.clone());
+
+                        Some(user)
+                    } else {
+                        return Err(ErrorResponse::new(ErrorResponseType::Forbidden, format!(
+                            "User with email '{}' already exists but is not linked to this provider.",
+                            user.email
+                        )));
+                    }
+                } else {
+                    None
+                }
+            }
+        };
+        debug!("user_opt:\n{:?}", user_opt);
+
+        let now = OffsetDateTime::now_utc().unix_timestamp();
+        let user = if let Some(mut user) = user_opt {
+            let mut old_email = None;
+            let mut forbidden_error = None;
+
+            if user.federation_uid.is_none()
+                || user.federation_uid.as_deref() != Some(&claims_user_id)
+            {
+                forbidden_error = Some("non-federated user or ID mismatch");
+            }
+
+            if user.auth_provider_id.as_deref() != Some("atproto") {
+                forbidden_error = Some("invalid login from wrong auth provider");
+            }
+
+            if let Some(err) = forbidden_error {
+                user.last_failed_login = Some(now);
+                user.failed_login_attempts =
+                    Some(user.failed_login_attempts.unwrap_or_default() + 1);
+                user.save(old_email).await?;
+
+                return Err(ErrorResponse::new(
+                    ErrorResponseType::Forbidden,
+                    err.to_string(),
+                ));
+            }
+
+            if Some(user.email.as_str()) != email.as_deref() {
+                old_email = Some(user.email);
+                user.email = email.as_ref().unwrap().to_string();
+            }
+
+            user.last_login = Some(now);
+            user.last_failed_login = None;
+            user.failed_login_attempts = None;
+
+            user.save(old_email).await?;
+            user
+        } else {
+            let new_user = User {
+                email: email.as_ref().unwrap().to_string(),
+                given_name: "N/A".to_string(),
+                family_name: None,
+                roles: Default::default(),
```

Comment:
For `roles` mapping, we should either have a proper `atproto` upstream provider configured, so they can be mapped, or add a config variable with default roles for atproto logins.

Most often, a downstream client will only accept logins if certain roles / groups match. Maybe @zicklag can provide information how they use this information internally.

Reply:
Sorry I didn't see this mention.

Right now the only thing we use roles for is to recognize admin users and give them access to the admin panel in Weird.

Actually I'm a little confused by the difference between roles and groups, but right now we only distinguish between users that can login at all, and users that are also admins.

Reply:
Depending on the setup, there is no difference, if you only take a single value into account. The roles + groups  become more powerful for complex situations when you combine them like `groups == engineers && roles == admin`.

@avdb13 if the typical `atproto` clients fits into the scheme of currently existing upstream IdP's, I would prefer the solution to add them there, but I guess they don't, because each user provides their own handle / callback, right?
In that case, blacklisting `atproto` as an upstream IdP name would be the best solution.